client.js

```diff
@@ -547,21 +559,51 @@ DHT.prototype._ongetpeers = function (query, peer) {
 }
 
 DHT.prototype._onannouncepeer = function (query, peer) {
+  var req = this._validatePeerUpdate(query, peer)
+
+  if (!req) {
+    return this._rpc.error(peer, query, [203, 'cannot `announce_peer` with bad token'])
+  }
+
+  var from = {host: req.host, port: peer.port}
+  var infoHash = query.a.info_hash
+
+  this.emit('announce_peer', infoHash, from)
+  this._addPeer(req, infoHash, from)
+  this._rpc.response(peer, query, {id: this._rpc.id})
+}
+
+DHT.prototype._onunannouncepeer = function (query, peer) {
+  var req = this._validatePeerUpdate(query, peer)
+
+  if (!req) {
+    return this._rpc.error(peer, query, [203, 'cannot `unannounce_peer` with bad token'])
+  }
+
+  var from = {host: req.host, port: peer.port}
+  var infoHash = query.a.info_hash
+
+  this.emit('unannounce_peer', infoHash, from)
+  this._removePeer(req, infoHash, from)
+  this._rpc.response(peer, query, {id: this._rpc.id})
+}
+
+DHT.prototype._validatePeerUpdate = function (query, peer) {
   var host = peer.address || peer.host
   var port = query.a.implied_port ? peer.port : query.a.port
   if (!port || typeof port !== 'number' || port <= 0 || port > 65535) return
   var infoHash = query.a.info_hash
   var token = query.a.token
   if (!infoHash || !token) return
 
-  if (!this._validateToken(host, token)) {
-    return this._rpc.error(peer, query, [203, 'cannot `announce_peer` with bad token'])
-  }
+  if (!this._validateToken(host, token)) return null
```

Comment:
Drop `null` to match code above?

Reply:
It still does the same. Kinda hard to see from the diff. Returning null in here triggers the .error handler in the caller